plutus-contract/src/Plutus/Contract/StateMachine.hs

```diff
@@ -128,8 +127,9 @@ getInput ::
     -> ChainIndexTx
     -> Maybe i
 getInput outRef tx = do
-    (_validator, Ledger.Redeemer r, _) <-
-      listToMaybe $ mapMaybe Tx.inScripts $ filter (\Tx.TxIn{Tx.txInRef} -> outRef == txInRef) $ Set.toList $ _citxInputs tx
+    -- We retrieve the correspondent redeemer according to the index of txIn in the list
+    let findRedeemer (ix, _) = Map.lookup (Tx.RedeemerPtr Tx.Spend ix) (_citxRedeemers tx)
+    Ledger.Redeemer r <- listToMaybe $ mapMaybe findRedeemer $ filter (\(_, Tx.TxIn{Tx.txInRef}) -> outRef == txInRef) $ zip [0..] $ _citxInputs tx
```

Comment:
Here is where we use the knowledge of inputs order to get the redeemer. I hope I got the idea correctly that index of `TxIn` is binded to the redeemer pointer.

Source: https://github.com/input-output-hk/cardano-node/issues/3389#issuecomment-1043165343

Maybe instead of relying on their order it's better to include the index. `Set TxIn ~> Set (TxInIndx, TxIn)`, that would make it more explicit.

Reply:
The fact that we now have a list instead of a set makes it clear enough that the order is significant in my opinion.